### ISSUE TITLE
[Crash] VB If operator causes ArgumentException: Delegate to an instance method cannot have null 'this' for delegates.

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#377]
Repro Steps: 
Write this line and run the project. 
```vb
Dim F As Func(Of Boolean) = If(True, Nothing, Function() True)
``` 
**Actual Results**: 

Exception is thrown:
``` 
System.ArgumentException: Delegate to an instance method cannot have null 'this'. 
at System.MulticastDelegate.ThrowNullThisInDelegateToInstance() 
```

**Expected Results**: 

Value of Nothing should be assigned to variable F.

[ported from TFS DevDiv 862907]
[Originally from Connect](https://connectadmin/Feedback/ConnectTab.aspx?FeedbackID=814131 "Connect Bug")

**Comment by gafter** [comment]
@AlekseyTs wrote:

The problem is that Anonymous Delegate type conversion doesn’t check whether the Anonymous Delegate instance is a null reference before creating a delegate of the target type by using its Invoke method. I changed the way conversion is lowered (shelveset Bug862907;alekseyt), it fixed the problem, but introduced another one.

Since now we capture the Anonymous Delegate instance in a local in order to use it twice (for null test and for target delegate creation), an Anonymous Delegate conversion is no longer supported in a lambda converted to an Expression Tree. Also, I believe there is a plan to implement a new feature in Dev14 to flow target type into conditional expression. That change will make the problem disappear in this scenario. Of course, one could come up with a code that explicitly assigns Nothing to a variable of an Anonymous Delegate type and then converts it to a different delegate type, but I believe a scenario like this would be rather pathological. Given that, my recommendation is to Won’t Fix this bug.

**Comment by gafter** [comment]
I would rather we fix the code gen bug when not in an expression tree than generate bad code in non-expression-tree scenario just because we have no way of generating the corresponding correct expression tree.

**Comment by gafter** [comment]
I'm licking this cookie.

**Comment by gafter** [comment]
Recommendation:

1. Fix the normal (non expression-tree) case.
2. Make sure the expression-tree case works the same as today (broken for Nothing).
3. File a new issue about getting the expression-tree case to behave the same as the normal case.

**Comment by gafter** [comment]
@AlekseyTs Please unassign if the rc2 timeframe is unrealistic for you.

**Comment by AlekseyTs** [comment]
This looks like a duplicate issue, the fix has been already implemented as 755ffcd0

**Comment by AlekseyTs** [comment]
@gafter, please verify that the issue has been addressed.
